cluster/cluster.go

```diff
@@ -155,13 +281,130 @@ func (p *Peer) warnIfAlone(logger log.Logger, d time.Duration) {
 		case <-p.stopc:
 			return
 		case <-tick.C:
-			if n := p.mlist.NumMembers(); n <= 1 {
-				level.Warn(logger).Log("NumMembers", n, "msg", "I appear to be alone in the cluster")
-			}
+			p.removeFailedPeers(timeout)
 		}
 	}
 }
 
+func (p *Peer) removeFailedPeers(timeout time.Duration) {
+	p.peerLock.Lock()
+	defer p.peerLock.Unlock()
+
+	now := time.Now()
+
+	keep := make([]peer, 0, len(p.failedPeers))
+	for _, pr := range p.failedPeers {
+		if pr.leaveTime.Add(timeout).After(now) {
+			keep = append(keep, pr)
+		} else {
+			level.Debug(p.logger).Log("msg", "failed peer has timed out", "peer", pr.Node, "addr", pr.Address())
+			delete(p.peers, pr.Name)
+		}
+	}
+
+	p.failedPeers = keep
+}
+
+func (p *Peer) handleReconnect(d time.Duration) {
+	tick := time.NewTicker(d)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-p.stopc:
+			return
+		case <-tick.C:
+			p.reconnect()
+		}
+	}
+}
+
+func (p *Peer) reconnect() {
+	p.peerLock.RLock()
+	failedPeers := p.failedPeers
+	p.peerLock.RUnlock()
+
+	logger := log.With(p.logger, "msg", "reconnect")
+	for _, pr := range failedPeers {
+		// No need to do book keeping on failedPeers here. If a
+		// reconnect is successful, they will be announced in
+		// peerJoin().
+		if _, err := p.mlist.Join([]string{pr.Address()}); err != nil {
+			p.failedReconnectionsCounter.Inc()
+			level.Debug(logger).Log("result", "failure", "peer", pr.Node, "addr", pr.Address())
+		} else {
+			p.reconnectionsCounter.Inc()
+			level.Debug(logger).Log("result", "success", "peer", pr.Node, "addr", pr.Address())
+		}
+	}
+}
+
+func (p *Peer) peerJoin(n *memberlist.Node) {
+	p.peerLock.Lock()
+	defer p.peerLock.Unlock()
+
+	var oldStatus PeerStatus
+	pr, ok := p.peers[n.Address()]
+	if !ok {
+		oldStatus = StatusNone
+		pr = peer{
+			status: StatusAlive,
+			Node:   n,
+		}
+	} else {
+		oldStatus = pr.status
+		pr.Node = n
+		pr.status = StatusAlive
+		pr.leaveTime = time.Time{}
+	}
+
+	p.peers[n.Address()] = pr
+	p.peerJoinCounter.Inc()
+
+	if oldStatus == StatusFailed {
+		level.Debug(p.logger).Log("msg", "peer rejoined", "peer", pr.Node)
+		p.failedPeers = removeOldPeer(p.failedPeers, pr.Name)
+	}
+}
+
+func (p *Peer) peerLeave(n *memberlist.Node) {
+	p.peerLock.Lock()
+	defer p.peerLock.Unlock()
+
+	pr, ok := p.peers[n.Address()]
+	if !ok {
+		// Why are we receiving a leave notification from a node that
+		// never joined?
+		return
+	}
+
+	pr.status = StatusFailed
+	pr.leaveTime = time.Now()
+	p.failedPeers = append(p.failedPeers, pr)
+	p.peers[n.Address()] = pr
+
+	p.peerLeaveCounter.Inc()
+	level.Debug(p.logger).Log("msg", "peer left", "peer", pr.Node)
+}
+
+func (p *Peer) peerUpdate(n *memberlist.Node) {
+	p.peerLock.Lock()
+	defer p.peerLock.Unlock()
+
+	pr, ok := p.peers[n.Address()]
+	if !ok {
+		// Why are we receiving an update from a node that never
+		// joined?
```

Comment:
Could be that the process has restarted and receives an out-of-bound notification?